scenes/controller/offline_controller/offline_controller.gd

```diff
@@ -0,0 +1,162 @@
+func _progress_worker_controller(efficiencies: Dictionary, cycles: int) -> Dictionary:
+	var total_efficiency: Dictionary = efficiencies["total_efficiency"]
+
+	# normalize "house -> worker" production
+	if total_efficiency.has("house"):
+		var worker_inc: int = Limits.safe_mult(
+			Game.params["house_workers"], total_efficiency.get("house", 0)
+		)
+		total_efficiency["worker"] = worker_inc
+	else:
+		total_efficiency["worker"] = 0
+
+	# check for decreasing resources
+	var ids: Array = total_efficiency.keys()
+	var decreasing_ids: Array = ids.filter(
+		func(id: String) -> bool: return total_efficiency[id] < 0
+	)
+	if !decreasing_ids.is_empty():
+		return {"decreasing_ids": decreasing_ids, "generated": {}}
+
+	# calculate production for cycles
+	var generated: Dictionary = {}
+	for id: String in ids:
+		var id_eff: int = total_efficiency[id]
+		var resource: ResourceGenerator = Resources.resource_generators[id]
+		if resource.is_dynamic_efficiency():
+			var de_id: String = resource.dynamic_efficiency_id
+			var de_id_eff: int = total_efficiency[de_id]
+			var generate: int = MathUtils.safe_dual_sum(id_eff, de_id_eff, cycles)
+			generated[id] = Limits.safe_add(generated.get(id, 0), generate)
+		else:
+			var generate: int = Limits.safe_mult(id_eff, cycles)
+			generated[id] = Limits.safe_add(generated.get(id, 0), generate)
+
+	return {"decreasing_ids": {}, "generated": generated}
+
+
+func _generate_resources(generated: Dictionary) -> bool:
+	var is_success: bool = false
+	for resource_id: String in generated:
+		var amount: int = generated[resource_id]
+		if amount == 0:
+			continue
+		is_success = true
+		SignalBus.resource_generated.emit(resource_id, amount, name)
+
+		var resource: ResourceGenerator = Resources.resource_generators[resource_id]
+		var worker_costs: Dictionary = resource.worker_costs
+		for worker_id: String in worker_costs:
+			var cost: int = Limits.safe_mult(amount, worker_costs[worker_id])
+			SignalBus.worker_generated.emit(worker_id, -cost, name)
+	return is_success
+
+
+func _progress_enemy_controller(
+	swordsman_storage: int, swordsman_eff: int, cycles: int, enemy_data: EnemyData
+) -> Dictionary:
+	var enemy_health: int = enemy_data.health_points
+	var overkill_factor: float = MathUtils.dual_sum_normalized(
+		swordsman_storage, swordsman_eff, cycles, enemy_health
+	)
+	var damage: int = MathUtils.safe_dual_sum(swordsman_storage, swordsman_eff, cycles)
+
+	var generated: Dictionary = {}
+	if enemy_data.is_last() && overkill_factor >= 1.0:
+		generated["soulstone"] = int(overkill_factor)
+	return {"overkill_factor": overkill_factor, "damage": damage, "generated": generated}
+
+
+func _trigger_enemy_controller(damage: int) -> void:
+	SignalBus.enemy_damage.emit(damage, name)
+
+
+##############
+## handlers ##
+##############
+
+
+func _handle_on_game_resumed(
+	seconds_delta: int, seconds_delta_expected: int, factor: float = 1.0
```

Comment:
Default is a factor of 1.0 (100%) efficiency, for now.

Related: https://trello.com/c/es3pEvUf/1-offline-progression#comment-667726dab1d01431ef17eba8